File: 算法分类/数组-字符串-链表/排序/Merge_Sort.py
def merge_sort(list):
    n = len(list)

    if n == 1:
        return list

    mid = n // 2

    left = merge_sort(list[:mid])
    right = merge_sort(list[mid:])

    left_cursor = 0
    right_cursor = 0
    result = []

    while left_cursor < len(left) and right_cursor < len(right):
        if left[left_cursor] <= right[right_cursor]:
            result.append(left[left_cursor])
            left_cursor += 1
        else:
            result.append(right[right_cursor])
            right_cursor += 1
    result += left[left_cursor:]
    result += right[right_cursor:]

    return result


items = [26, 45, 44, 6, 9, 54, 93]

result = merge_sort(items)

print(result)

# 最好情况: O(nlogn)
# 最坏情况: O（nlogn）
# 平均情况: O(nlogn)
# 稳定性: 稳定
# 辅助空间: O(n)

File: Templates(Swift)/Templates/Sort/MergeSort.swift
//
//  MergeSort.swift
//  Templates
//
//  Created by Harry Yan on 1/10/22.
//

import Foundation

func mergeSort(_ nums: [Int], _ start: Int, _ end: Int) -> [Int] {
    guard start < end else { return nums }
    
    let mid = start + (end - start) / 2
    let left = mergeSort(nums, start, mid)
    let right = mergeSort(nums, mid+1, end)
    
    var sortedList: [Int] = []
    var p1 = 0
    var p2 = 0
    
    while p1 < left.count || p2 < right.count {
        if p1 == left.count {
            sortedList.append(right[p2])
            p2 += 1
        } else if p2 == right.count {
            sortedList.append(right[p1])
            p1 += 1
        } else {
            if nums[p1] < nums[p2] {
                sortedList.append(right[p1])
                p1 += 1
            } else {
                sortedList.append(right[p2])
                p2 += 1
            }
        }
    }
    
    return sortedList
}
